### ISSUE TITLE
Fixing mod deployment folder

### ISSUE THREAD

**Issue by taosecurity** [limo-app/limo#54]
Hello,

Thank you for your work on this neat too. Apologies if PEBCAK. 

I was trying to install a Starfield mod, Ryujin Shipping Services:

https://www.nexusmods.com/starfield/mods/12049

I noticed that the esm 

VA Ryujin Shipping Services.esm 

is packaged within the directory

VA Ryujin Shipping Services

rather than the more common Data directory.

As a result, when I use Limo to deploy this mod, it puts VA Ryujin Shipping Services.esm in VA Ryujin Shipping Services. Starfield can't find it.

When I use Vortex on Windows, Vortex ignores the VA Ryujin Shipping Services directory and places VA Ryujin Shipping Services.esm in Data like other .esm.

I can't see a way to tell Limo to put VA Ryujin Shipping Services.esm in Data and not VA Ryujin Shipping Services.

Is there a way to do this? Thanks again.

Richard

**Comment by limo-app** [comment]
Hi,
Assuming you have a deployer pointing to the `data` directory of Starfield:
The *Install Mod* dialog has a setting called *Root level*. This setting allows you to "cut off" parts of directory tree of the mod you are currently installing. Setting this to 1 will remove the `VA Ryujin Shipping Services` directory, ensuring that its contents (including VA `Ryujin Shipping Services.esm`) will be placed inside the `data` directory when clicking *Deploy*.

For more details, refer to the [Installing mods section of the wiki](https://github.com/limo-app/limo/wiki#installing-mods).

**Comment by taosecurity** [comment]
Ah, very good! I will give this a try. Thank you very much!

**Comment by taosecurity** [comment]
Hello again,

I tried setting Root Level to 1. I ended up with a link in 

~/.steam/debian-installation/steamapps/common/Starfield

pointing to the esm in my staging directory.

lrwxrwxrwx 1 rb rb 64 Nov 10 08:51 'VA Ryujin Shipping Services.esm' -> '/home/rb/Downloads/LimoStaging/2/VA Ryujin Shipping Services.esm'

I think because the Install deployer points to

~/.steam/debian-installation/steamapps/common/Starfield

and the other mods I have tested have a Data directory, eg End-to-End Mission Boards

https://www.nexusmods.com/starfield/mods/10390

mods like that put the .esm in the Data directory as desired with Root as 0.

Does this mean I should edit the Install deployer to point to

/home/rb/.steam/debian-installation/steamapps/common/Starfield/Data 

and set the Root for all mods I deploy as 1?

Sincerely,

Richard

**Comment by taosecurity** [comment]
Apologies for the quick follow-up... I tested what I just said and I got this error:

Could not rename file '/home/rb/Downloads/LimoStaging/0/VA Ryujin Shipping Services.esm' to '/home/rb/Downloads/LimoStaging/0/VA Ryujin Shipping Services.esm' because the target already exists

I did end up with the link I expected though in Data:

~/.steam/debian-installation/steamapps/common/Starfield/Data$ ls -al VA*
lrwxrwxrwx 1 rb rb 64 Nov 10 09:14 'VA Ryujin Shipping Services.esm' -> '/home/rb/Downloads/LimoStaging/0/VA Ryujin Shipping Services.esm'

However, I did not have a problem with End-to-End Mission Boards.

Sincerely,

Richard

**Comment by limo-app** [comment]
> Does this mean I should edit the Install deployer to point to
> 
> /home/rb/.steam/debian-installation/steamapps/common/Starfield/Data
> 
> and set the Root for all mods I deploy as 1?

I do not own Starfield so i don't know how those mods usually work. But from my experience with Skyrim, it is best to have one deployer pointing to `Starfield/Data` and one pointing to `Starfield`. Almost all mods will end up in the Data deployer (with adjusted root level). This makes it easier to deal with cases where the mod author did not include a Data directory in their archive.


> Could not rename file '/home/rb/Downloads/LimoStaging/0/VA Ryujin Shipping Services.esm' to '/home/rb/Downloads/LimoStaging/0/VA Ryujin Shipping Services.esm' because the target already exists

This happens when a case matching deployer tries to adapt a file name and the adapted name already exists. Does this happen every time you try to deploy mods? Did you manually change any files for this mod?

**Comment by limo-app** [comment]
To clarify the deployers: Create one pointing to `Starfield`, let's call that the *Bin* deployer and another one pointing to `Starfield/Data`, let's call that the *Data* deployer. Choosing which deployer manages a mod should depend on where that mod's files should go, not how it is packaged. For Bethesda games, almost all mods will go into the `Data` directory, unless the mod other says otherwise.
When installing a mod, you can select which deployers should manage it (bottom right of the installer). Almost all mods you are going to install should use the *Data* deployer. If a mod should be managed by *Bin*, the mod author usually says so in bold red letters on the mod page, since those mods cannot be installed with most other mod managers. Even if a mod contains a *Data* directory, i recommend you simply change the root level accordingly and add it to the *Data* deployer. By default, all deployers will deploy their mods one after the other when you click deploy.

As for the error: This might be a bug. Could you please send me the `lmm_mods.json` file in your staging directory (`/home/rb/Downloads/LimoStaging/`) as well as the file names in in the mod's directory (`'/home/rb/Downloads/LimoStaging/0`) and the file names in Starfield's *Data* directory (just the output of `ls`). I am trying to understand why the file renaming fails.

**Comment by ActuallyTheOwner** [comment]
I was unable to build loot/libloot on debian or gentoo as both distros now use GCC 13 and not 10, it seems dependencies are based on Ubuntu which nolonger reflects most distros; is this related?

**Comment by limo-app** [comment]
> I was unable to build loot/libloot on debian or gentoo as both distros now use GCC 13 and not 10, it seems dependencies are based on Ubuntu which nolonger reflects most distros; is this related?

You are correct that dependencies are based on Ubuntu, as I only tested building on Ubuntu personally, though i know from others that it works on Arch as well. I used GCC 13 to build Limo in the past without issues and now GCC 14.2. Limo is using some C++ 23 features so older compilers may not work.
